File: task_sheets/sheet3/src/SinglePoint.java
public class SinglePoint {
    /*
    Returns the number of binary digits that can be printed from single point percision.
     */
    public static float calcSinglePoint(){
        float one = 1;
        float seps = 1;
        float appone = one + seps;
        for( int ipow = 1; ipow <= 1000; ipow++){

            seps = seps / 2;
            appone = 1 + seps;
            if (Math.abs(appone) == 1){
                return ipow;
            }
        }
        System.out.print("Did not make it after 1000 iterations");
        return 0;
    }

    public static void main(String[] args){
        float numBinaryDigits = calcSinglePoint();
        System.out.println("The number of binary digits that can be printed is " + calcSinglePoint());
        System.out.println("The smallest number this corresponds to is " + Math.pow(2,-numBinaryDigits));
    }
}

File: lib/SinglePoint.py
from numpy import float32

def singlePoint():
    one = float32(1.0)
    seps = float32(1.0)
    appone = float32(1.0)
    for ipow in range(1, 1000):
        seps = seps / 2
        seps = float32(seps)
        appone = one + seps
        if (abs(appone) == 1):
            return ipow
    print("Didn't make it after 1000 Iterations")


if __name__ == "__main__":
    print(f"The number of binary digits that can be printed is {singlePoint()}")
    print(f"The smallest number this corresponds to is {2 ** -singlePoint()}")